### ISSUE TITLE
Difference between daemon and normal execution ?

### ISSUE THREAD

**Issue by yan-sag** [savonet/liquidsoap#3450]
Hi,
I'm trying to migrate from v1.4.4 to v2.1.2. I used to install from .deb packages, but I intalled 2.1.2 with opam.

My script runs well when I use `liquidsoap ./encoder5.liq`, but it fails when I try to check it with `liquidsoap --check encoder5.liq` (segmentation fault), or when I 'daemonize' it with the liquidsoap-daemonize.sh script and start the service (seg fault too).

Is there a difference between thoses execution types ?

My script :

```liquidsoap
#!/usr/bin/liquidsoap

log.file.path.set("/home/user/logs/encoder5.log")
log.level.set(5)

settings.frame.audio.samplerate.set(48000)
settings.frame.audio.channels.set(2)

def crossfade(a,b)
  add(normalize=false,
      [ sequence([ blank(duration=0.5),
                   fade.initial(duration=1.,b) ]),
        fade.final(duration=1.,a) ])
end

playlist = playlist('/home/user/musique/')
playlist = mksafe(id="plsafe",playlist)

input_jack = input.jack(clock_safe=false)
input_jack = mksafe(id="insafe", input_jack)


radio = input_jack
radio = fallback( track_sensitive=false,
		  [ blank.strip(max_blank=60., radio), playlist] )

output.file(%wav(stereo=true, channels=2, samplesize=16, header=true, duration=3600.),
	    '/home/user/recorded/%Y-%m-%d/%Y-%m-%d-%H_%M_%S.wav',
	    reopen_when={0m0s},
	    radio
	   )

opus_stream = mksafe(buffer.adaptative(limit=1.02, reset=true, radio))
clock.assign_new(id="opus_clk", [opus_stream])
output.icecast(
    %opus(channels=2),
    host="127.0.0.1",
    port=8000,
    password="***",
    mount="primitive.opus",
    opus_stream
)

mp3_stream = mksafe(buffer(radio))
clock.assign_new(id="mp3_clock", [mp3_stream])
output.icecast(
    %mp3(bitrate=160, stereo=true),
    host="127.0.0.1",
    port=8000,
    password="***",
    mount="primitive.mp3",
    mp3_stream
)
```

**Comment by toots** [comment]
Hi @yan-sag. This looks like a bug. What version are you using? I've run your script locally but it seems to be using an old API for fades.

**Comment by yan-sag** [comment]
Hi @toots,
The script launched by `liquidsoap ./encoder5.liq` crashed  this night: the log stops abruptly.

The daemonized encoder5 crashes just after starting, in journalctl :
`encoder kernel: liquidsoap[73846]: segfault at 563762811b4d ip 00007f2688c08bc1 sp 00007ffc1f866f60 error 4 in libsrt.so.1.5.1[7f2688b88000+98000] likely on CPU 2 (core 2, socket 0)`
`encoder kernel: Code: 10 48 85 db 0f 84 8f 00 00 00 48 8b 7b 18 e8 26 ff ff ff 80 7b 48 00 48 8b 6b 10 74 3e 4c 8b 63 40 4d 85 e4 74 35 49 8b 04 24 <48> 8b 40 28 48 3b 05 84 93 03 00 75 6a 48>`
`encoder systemd[1]: encoder5-daemon-liquidsoap.service: Control process exited, code=killed, status=11/SEGV`

here's my config :
Debian 12
Opam 2.1.2
Ocaml 4.13.1
Liquidsoap 2.1.2 (I can't install newer one)

I'm installing a new machine with fresh debian and fresh opam to try.

**Comment by yan-sag** [comment]
I think my LS install is broken :

> `$ liquidsoap --version`
> Liquidsoap 2.1.2
> Copyright (c) 2003-2022 Savonet team
> Liquidsoap is open-source software, released under GNU General Public License.
> See <http://liquidsoap.info> for more information.
> Erreur de segmentation

The seg fault is due to liquidsoap (or something else), not my script..

**Comment by toots** [comment]
Indeed. Have you ever used `gdb`? You should be able to do:
```
% gdb `which liquidsoap`
> r --version
.. sefgault
> thread apply all bt
```

**Comment by yan-sag** [comment]
Hi @toots ,
I face the same problem with my fresh install :
- debian 12
- ocaml 4.13.1
- opam 2.1.2
- liquidsoap 2.1.4

I build liquidsoap with
`opam install liquidsoap ffmpeg opus srt bjack alsa taglib mad lame vorbis cry ssl samplerate magic ocurl`

Here' the gdb trace :
```
(gdb) r --version
Starting program: /home/primitive/.opam/default/bin/liquidsoap --version
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffe60966c0 (LWP 130628)]
[New Thread 0x7fffe58956c0 (LWP 130629)]
[New Thread 0x7fffdffff6c0 (LWP 130630)]
Liquidsoap 2.1.4
Copyright (c) 2003-2022 Savonet team
Liquidsoap is open-source software, released under GNU General Public License.
See <http://liquidsoap.info> for more information.
[Thread 0x7fffdffff6c0 (LWP 130630) exited]

Thread 1 "liquidsoap" received signal SIGSEGV, Segmentation fault.
0x00007ffff7f6faf1 in std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, srt::PacketFilter::ManagedPtr, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, srt::PacketFilter::ManagedPtr> > >::~map() ()
   from /usr/local/lib/libsrt.so.1.5
```
```
(gdb) thread apply all bt

Thread 3 (Thread 0x7fffe58956c0 (LWP 130629) "liquidsoap"):
#0  0x00007ffff451d7dc in __GI___select (nfds=0, readfds=0x0, writefds=0x0, exceptfds=0x0, timeout=0x7fffe5894230) at ../sysdeps/unix/sysv/linux/select.c:69
#1  0x0000555555f49ef4 in caml_thread_tick ()
#2  0x00007ffff44a8044 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#3  0x00007ffff452861c in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

Thread 2 (Thread 0x7fffe60966c0 (LWP 130628) "liquidsoap"):
#0  __futex_abstimed_wait_common64 (private=0, cancel=true, abstime=0x0, op=393, expected=0, futex_word=0x5555565558c8 <log_condition+40>) at ./nptl/futex-internal.c:57
#1  __futex_abstimed_wait_common (futex_word=futex_word@entry=0x5555565558c8 <log_condition+40>, expected=expected@entry=0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=private@entry=0, cancel=cancel@entry=true) at ./nptl/futex-internal.c:87
#2  0x00007ffff44a4e0b in __GI___futex_abstimed_wait_cancelable64 (futex_word=futex_word@entry=0x5555565558c8 <log_condition+40>, expected=expected@entry=0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=private@entry=0) at ./nptl/futex-internal.c:139
#3  0x00007ffff44a7468 in __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x555556555860 <log_mutex>, cond=0x5555565558a0 <log_condition>) at ./nptl/pthread_cond_wait.c:503
#4  ___pthread_cond_wait (cond=cond@entry=0x5555565558a0 <log_condition>, mutex=mutex@entry=0x555556555860 <log_mutex>) at ./nptl/pthread_cond_wait.c:618
#5  0x0000555555ef414b in ocaml_srt_process_log (cb=<optimized out>) at srt_generated_stubs.c:36
#6  0x0000555555d0ff47 in camlSrt__fun_3401 () at src/srt.ml:462
#7  0x0000555555e8dd75 in camlThread__fun_314 () at thread.ml:41
#8  0x0000555555f74b89 in caml_start_program ()
#9  0x0000555555f6a83d in caml_callback_exn ()
#10 0x0000555555f4abf0 in caml_thread_start ()
#11 0x00007ffff44a8044 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#12 0x00007ffff452861c in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

Thread 1 (Thread 0x7fffe6723340 (LWP 130625) "liquidsoap"):
#0  0x00007ffff7f6faf1 in std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, srt::PacketFilter::ManagedPtr, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, srt::PacketFilter::ManagedPtr> > >::~map() () at /usr/local/lib/libsrt.so.1.5
#1  0x00007ffff445cf77 in __cxa_finalize (d=0x7ffff18b3000) at ./stdlib/cxa_finalize.c:83
#2  0x00007ffff180de47 in  () at /lib/x86_64-linux-gnu/libsrt-gnutls.so.1.5
#3  0x00007ffff2f62530 in  ()
#4  0x00007ffff7fcc12a in _dl_call_fini (closure_map=0x7fffffffde40, closure_map@entry=0x7ffff2f62530) at ./elf/dl-call_fini.c:43
#5  0x00007ffff7fcf764 in _dl_fini () at ./elf/dl-fini.c:78
#6  0x00007ffff445d4dd in __run_exit_handlers (status=0, listp=0x7ffff45f1820 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at ./stdlib/exit.c:116
#7  0x00007ffff445d61a in __GI_exit (status=<optimized out>) at ./stdlib/exit.c:146
#8  0x0000555555f6572e in caml_do_exit ()
#9  0x0000555555f6592c in  ()
#10 0x0000555555e97ce3 in camlStdlib__exit_476 () at stdlib.ml:570
#11 0x0000555555ecccec in camlStdlib__Arg__parse_and_expand_argv_dynamic_aux_379 () at arg.ml:278
#12 0x0000555555a13a2a in camlMain__parse_1100 () at main.ml:381
#13 0x0000555555a13cdb in camlMain__fun_2315 () at main.ml:432
#14 0x0000555555e9a7f6 in camlStdlib__List__iter_261 () at list.ml:110
#15 0x0000555555e46eec in camlDtools__Dtools_impl__exec_810 () at src/dtools_impl.ml:441
#16 0x0000555555e9a7f6 in camlStdlib__List__iter_261 () at list.ml:110
#17 0x0000555555e46f29 in camlDtools__Dtools_impl__exec_810 () at src/dtools_impl.ml:444
#18 0x0000555555a12653 in camlRunner__entry () at main.ml:572
#19 0x0000555555a07f99 in caml_program ()
#20 0x0000555555f74b89 in caml_start_program ()
#21 0x0000555555f53854 in caml_startup_common ()
#22 0x0000555555f538cb in caml_main ()
#23 0x0000555555a05a0c in main ()
(gdb)
```

On my previous test, it was just after trying to build with SRT support that the segfaults occured.

**Comment by toots** [comment]
Thanks for getting those this is very helpful. It looks like SRT is initialized before the daemonization fork, which creates issues.

#3473 should address that, it would be nice is y'all could test it. The CI runs should produce assets for testing.

This being said, the feature to daemonize from inside a liquidsoap script is deprecated now. It is strongly recommended to use a modern daemonization facility such as `systemd`.

The reason is because script evaluation are complex and it's nearly impossible for us to prevent improper resource initalization before forking and changing to the user/group specified for the daemon.

This is because we do need to parse the whole script to evaluate the daemonization setting. Meanwhile, there are a lot of other operations that can happen at script evaluation that can create resources such as sockets and etc. that, if created before daemonization would potentially inherit the initial environment/user/group context. Also, these resources are shared between the forked daemon process and the parent (terminated) process, which can create issues such as the one here I believe.

TL;DR: use `systemd` or equivalent, they can deal with all these daemonization details _before_ we eve get to start the application!

**Comment by yan-sag** [comment]
Hi @toots,
Thanks for your detailled reply and your commits.

I git-checked-out the branch you merged, opam-pined it and successfully build liquidsoap 2.2.2.
The segfault due to SRT does'nt happen then, except if I ctrl+c the process.
